File: backend/app/api/v1/replenishment.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_role
from app.core.warehouse.replenishment import ReplenishmentService
from app.models.warehouse import StorageLocation
from app.schemas.replenishment import ReplenishmentRuleUpdate, ReplenishmentTaskOut

router = APIRouter(prefix="/api/v1/replenishment", tags=["Replenishment"])


@router.get("/tasks", response_model=list[ReplenishmentTaskOut])
def list_tasks(db: Session = Depends(get_db)):
    return ReplenishmentService(db).list()


@router.post("/generate", response_model=list[ReplenishmentTaskOut])
def generate_tasks(
    db: Session = Depends(get_db),
    user: dict = Depends(require_role("admin", "supervisor")),
):
    return ReplenishmentService(db).generate(user["username"])


@router.put("/locations/{location_id}/rule")
def configure_rule(
    location_id: int,
    request: ReplenishmentRuleUpdate,
    db: Session = Depends(get_db),
    _user: dict = Depends(require_role("admin")),
):
    if request.maximumQty <= request.minimumQty:
        raise HTTPException(
            status_code=400, detail="Maximum quantity must exceed minimum quantity"
        )
    location = (
        db.query(StorageLocation)
        .filter(StorageLocation.location_id == location_id)
        .first()
    )
    if not location:
        raise HTTPException(status_code=404, detail="Storage location not found")
    location.replenishment_sku = request.internalSku
    location.replenishment_min_qty = request.minimumQty
    location.replenishment_max_qty = request.maximumQty
    db.commit()
    return {"success": True}


@router.post("/tasks/{task_id}/complete", response_model=ReplenishmentTaskOut)
def complete_task(
    task_id: int,
    db: Session = Depends(get_db),
    user: dict = Depends(require_role("admin", "supervisor", "operator")),
):
    try:
        return ReplenishmentService(db).complete(task_id, user["username"])
    except ValueError as exc:
        raise HTTPException(
            status_code=404 if "not found" in str(exc).lower() else 409, detail=str(exc)
        ) from exc

File: frontend/src/app/api/replenishment.ts
import axios from 'axios';

const API_BASE_URL = '/api/v1';

export interface ReplenishmentTask {
  taskId: number;
  internalSku: string;
  lotId: number;
  fromLocationId: number;
  toLocationId: number;
  quantity: number;
  status: string;
}

export interface ReplenishmentRuleUpdate {
  internalSku: string;
  minimumQty: number;
  maximumQty: number;
}

export async function listReplenishmentTasks(): Promise<ReplenishmentTask[]> {
  const response = await axios.get<ReplenishmentTask[]>(`${API_BASE_URL}/replenishment/tasks`);
  return response.data;
}

export async function generateReplenishmentTasks(): Promise<ReplenishmentTask[]> {
  const response = await axios.post<ReplenishmentTask[]>(`${API_BASE_URL}/replenishment/generate`);
  return response.data;
}

export async function completeReplenishmentTask(taskId: number): Promise<ReplenishmentTask> {
  const response = await axios.post<ReplenishmentTask>(
    `${API_BASE_URL}/replenishment/tasks/${taskId}/complete`
  );
  return response.data;
}

export async function updateReplenishmentRule(
  locationId: number,
  rule: ReplenishmentRuleUpdate
): Promise<{ success: boolean }> {
  const response = await axios.put(`${API_BASE_URL}/replenishment/locations/${locationId}/rule`, rule);
  return response.data;
}

export async function listStorageLocations(): Promise<
  Array<{ locationId: number; locationCode: string }>
> {
  const response = await axios.get(`${API_BASE_URL}/cycle-counts/locations`);
  return response.data;
}

export async function listItemsForDropdown(): Promise<
  Array<{ internalSku: string; description: string }>
> {
  const response = await axios.get(`${API_BASE_URL}/purchase-orders/items`);
  return response.data;
}
